module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,713 @@
+
+#define	BENCH_RES_FMT	"RAIDZ_math: [%-7s] %-8s %5ld.%.3ld MB/s\n"
+#define	BENCH_SKIP_FMT	"RAIDZ_math: [%-7s] is not supported.\n"
+
+#if defined(_KERNEL)
+#define	bench_print(x...)	cmn_err(CE_CONT, x)
+#else
+#define	bench_print(x...)	do { } while (0)
+#endif
+
+void
+vdev_raidz_math_init(void)
+{
+	const int impl_num = ARRAY_SIZE(raidz_all_maths);
+	const int tgtidx[3] = {3, 4, 5};
+	raidz_math_ops_t const *curr_impl;
+	unsigned long long run_count, best_run_count, speed;
+	hrtime_t t_start;
+	zio_t *bench_zio = NULL;
+	raidz_map_t *bench_rm = NULL;
+	uint64_t bench_parity;
+	int i, c, fn, impl;
+	int bench_rec_orig[7][3] = {
+		{1, 2, 3}, {0, 2, 3},
+		{0, 1, 3}, {2, 3, 4},
+		{1, 3, 4}, {0, 3, 4},
+		{3, 4, 5}
+	};
+
+	/* init & vdev_raidz_impl_lock */
+	rw_init(&vdev_raidz_impl_lock, NULL, RW_DEFAULT, NULL);
+
+#if !defined(_KERNEL)
+	/*
+	 * It is possible that this routine is called multiple times
+	 * inside user-space test. Check if we already initialized the
+	 * fastest ops and skip the benchmark.
+	 */
+	if (raidz_math_initialized) {
+		goto init_done;
+	}
+#endif
+
+	/* init scalar multiplication lookup table */
+	raidz_init_scalar_mul_lt();
+
+	/* move supported maths into raidz_supp_maths */
+	for (i = 0, c = 0; i < impl_num; i++) {
+		curr_impl = raidz_all_maths[i];
+
+		/* init kstat */
+		init_raidz_kstat(raidz_math_kstats + i, curr_impl->name);
+
+		if (curr_impl->is_supported()) {
+			raidz_supp_maths[c++] = (raidz_math_ops_t *) curr_impl;
+		} else {
+			bench_print(BENCH_SKIP_FMT, curr_impl->name);
+			continue;
+		}
+	}
+	raidz_supp_maths_cnt = c;	/* number of supported algos */
+
+	/* init  kstat for original algo  routines */
+	init_raidz_kstat(&(raidz_math_kstats[impl_num]), "original");
+
+	/* Fake an zio and run the benchmark on it */
+	bench_zio = kmem_alloc(sizeof (zio_t), KM_SLEEP);
+	VERIFY(bench_zio);
+	bzero(bench_zio, sizeof (zio_t));
+
+	bench_zio->io_offset = 0;
+	bench_zio->io_size = BENCH_ZIO_SIZE; /* only data columns */
+	bench_zio->io_data = zio_data_buf_alloc(BENCH_ZIO_SIZE);
+	if (!bench_zio->io_data) {
+		cmn_err(CE_WARN, "RAIDZ: Can not allocate benchmark memory!");
+		/* use last supported math as the fastest */
+		cmn_err(CE_WARN, "RAIDZ: Using [%s] implementation",
+			raidz_supp_maths[raidz_supp_maths_cnt-1]->name);
+		memcpy(&vdev_raidz_fastest_impl,
+			raidz_supp_maths[raidz_supp_maths_cnt-1],
+			sizeof (raidz_math_ops_t));
+
+		kmem_free(bench_zio, sizeof (zio_t));
+		return;
+	}
+
+	/*
+	 * Benchmark new RAIDZ implementations
+	 */
+
+	/* Find the fastest gen functions */
+	for (fn = 0; fn < RAIDZ_GEN_NUM; fn++) {
+		bench_parity = fn+1;
+		best_run_count = 0;
+
+		/* New raidz_map is needed for rach generate_p/q/r */
+		bench_rm = vdev_raidz_map_alloc(bench_zio, 9,
+		    BENCH_D_COLS+bench_parity, bench_parity);
+
+		for (impl = 0; impl < impl_num; impl++) {
+			curr_impl = raidz_all_maths[impl];
+
+			if (!curr_impl->is_supported()) {
+				continue;
+			}
+
+			kpreempt_disable();
```

Comment:
It's unsafe to disable preemption for these benchmarks because internally these tests cab perform sleeping allocations.  The kernel considers this to be an atomic context which triggers the usual BUG.  Since there isn't an easy way to make these allocations atomic I suggest dropping the `kpreempt_disable/enable()` and we'll have to live with the potentially precise benchmark results.

```
BUG: sleeping function called from invalid context at mm/slub.c:941
in_atomic(): 1, irqs_disabled(): 0, pid: 564, name: systemd-modules
```